clippy_lints/src/pointer_in_nomem_asm_block.rs

```diff
@@ -0,0 +1,84 @@
+use clippy_utils::diagnostics::span_lint;
+use rustc_ast::InlineAsmOptions;
+use rustc_hir::{Expr, ExprKind, InlineAsm, InlineAsmOperand};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::Span;
+
+fn has_in_operand_pointer(cx: &LateContext<'_>, asm_op: &InlineAsmOperand<'_>) -> bool {
+    let asm_in_expr = match asm_op {
+        InlineAsmOperand::SymStatic { .. }
+        | InlineAsmOperand::Out { .. }
+        | InlineAsmOperand::Const { .. }
+        | InlineAsmOperand::SymFn { .. }
+        | InlineAsmOperand::Label { .. } => return false,
+        InlineAsmOperand::SplitInOut { in_expr, .. } => in_expr,
+        InlineAsmOperand::In { expr, .. } | InlineAsmOperand::InOut { expr, .. } => expr,
+    };
+
+    // This checks for raw ptrs, refs and function pointers - the last one
+    // also technically counts as reading memory.
+    return cx.typeck_results().expr_ty(asm_in_expr).is_any_ptr();
+}
+
+fn yeet_lint(cx: &LateContext<'_>, op_spans: Vec<Span>) {
+    span_lint(
+        cx,
+        POINTER_IN_NOMEM_ASM_BLOCK,
+        op_spans,
+        "passing pointer to nomem asm block",
```

Comment:
Yeah, InlineAsm here doesn't have the spans, but it could be passed from earlier staged of compilation. Could we implement this in rustc, even though it will be for some time only used in clippy?

Reply:
If you want to do that, sure. This has been done a few times before, e.g. https://github.com/rust-lang/rust/pull/117585, although that one was required for the lint, whereas this seems more like a nice to have. I suppose it doesn't hurt if it doesn't regress anything.

That said, I think just simple note/help messages even w/o spans also work just fine here.